### PR TITLE
feat: Fixed metric calculation and made all grpo metrics token-level

### PR DESCRIPTION
**Pull request by SahilJain314**
I noticed that the loss_type of the GRPO loss was impacting how metrics were calculated. Changed all losses to always accept both global token and sequence counts so that metrics can be calculated based on whichever they should be. Also fixed a merge artifact that was causing sampling_importance_weights to be miscalculated.

### CODE REVIEW

File: nemo_rl/algorithms/loss_functions.py [hunk 1]
```diff
@@ -124,25 +125,20 @@ def __call__(
         reference_policy_logprobs = data["reference_policy_logprobs"][:, 1:]
 
         mask = token_mask * sample_mask.unsqueeze(-1)
+        if global_valid_seqs is None:
```
Comment (ashors1) [inline nemo_rl/algorithms/loss_functions.py]: Can we require that users pass this in rather than setting a default? This will not be correct if num_microbatches > 1

File: nemo_rl/algorithms/loss_functions.py [hunk 2]
```diff
@@ -315,6 +310,10 @@ def __call__(
         token_mask = data["token_mask"][:, 1:]
         sample_mask = data["sample_mask"]
         mask = token_mask * sample_mask.unsqueeze(-1)
+        if global_valid_seqs is None:
```
Comment (ashors1) [inline nemo_rl/algorithms/loss_functions.py]: unused

File: nemo_rl/algorithms/loss_functions.py (hunk 1 above)
Comment (SahilJain314) [reply nemo_rl/algorithms/loss_functions.py]: done

File: nemo_rl/algorithms/loss_functions.py (hunk 2 above)
Comment (SahilJain314) [reply nemo_rl/algorithms/loss_functions.py]: removed, thanks

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 3]
```diff
@@ -312,24 +312,29 @@ def train(
...
+                        "token_mask must be present in the data when using token-level loss"
                     )
+                    norm_factor = global_valid_toks.item()
                 elif loss_fn.loss_type == LossType.SEQUENCE_LEVEL:
-                    ## get number of valid samples in the global batch
-                    total_valid_tokens_or_seqs = torch.sum(global_batch["sample_mask"])
-                    torch.distributed.all_reduce(
-                        total_valid_tokens_or_seqs, group=self.dp_mesh.get_group()
-                    )
+                    norm_factor = global_valid_seqs.item()
```
Comment (ashors1) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: WDYT about removing this? If we get rid of the `norm_factor` (which is currently only used for logging), we also loosen the restriction that every loss function needs an associated `LossType`, which might make things simpler for users looking to add a custom loss function.

We could just log both `global_valid_toks` and `global_valid_seqs` instead

File: nemo_rl/algorithms/interfaces.py [hunk 4]
```diff
@@ -40,6 +44,16 @@ def __call__(
...
+            global_valid_seqs: Optional[torch.Tensor]
+                If provided, this tensor should contain the number of valid sequences in the microbatch.
+                If not provided, it will be computed as the sum of the sample_mask in the data dict.
+                This is used for global normalization for losses/metrics that are computed at the sequence level
+                and need to be aggregated across all microbatches.
+            global_valid_toks: Optional[torch.Tensor]
+                If provided, this tensor should contain the number of valid tokens in the microbatch.
+                If not provided, it will be computed as the sum of the token_mask in the data dict.
+                This is used for global normalization for losses/metrics that are computed at the token level
+                and need to be aggregated across all microbatches.
```
Comment (parthchadha) [inline nemo_rl/algorithms/interfaces.py]: Are these optional? I don't think we are computing the default value in `ClippedPGLossFn`?

Comment (SahilJain314) [reply nemo_rl/algorithms/interfaces.py]: good catch, fixed.

File: nemo_rl/models/policy/dtensor_policy_worker.py (hunk 3 above)
Comment (SahilJain314) [reply nemo_rl/models/policy/dtensor_policy_worker.py]: That makes sense to me, done .
